[WIP]: Sheldon diagnostics

vkatsuba opened the pull request.

vkatsuba commented on rebar.config:
```diff
@@ -17,6 +17,7 @@
        , {tdiff,         "0.1.2"}
        , {uuid,          "2.0.1", {pkg, uuid_erl}}
        , {gradualizer, {git, "https://github.com/josefs/Gradualizer.git", {ref, "e93db1c"}}}
+       , {rebar3_sheldon, "0.4.1"}
```
The `sheldon`works only with  plain-text - at the same time `rebar3_sheldon` was created and adapted specifically for the Erlang code.

vkatsuba commented on apps/els_lsp/src/els_diagnostics.erl:
```diff
@@ -146,3 +149,12 @@ valid(Ids0) ->
   Valid.
+
+-spec extra(list()) -> ok.
+extra([]) ->
```
unfortunately, I do not know this project so deeply and was guided by the already created diagnostics

vkatsuba replied:
Unfortunately, I do not know this project deeply and decided to use this approach. Also I was think about extra handling of enabled diagnostics, eg: if additional manipulation with the enabled diagnostics is needed, we can to expand this function.

vkatsuba commented on apps/els_lsp/src/els_sheldon_diagnostics.erl:
```diff
@@ -0,0 +1,114 @@
+diagnostic(Msg, Ln, Severity) ->
+  Range   = els_protocol:range(#{from => {Ln, 1}, to => {Ln + 1, 1}}),
+  Message = els_utils:to_binary(Msg),
+  [#{ range    => Range
```
unfortunately, I do not know this project so deeply and was guided by the already created diagnostics

vkatsuba replied:
fixed

vkatsuba commented on apps/els_lsp/src/els_sheldon_diagnostics.erl:
```diff
@@ -0,0 +1,114 @@
+      of
+        [] -> [];
+        Problems -> format_diagnostics(Problems)
+      catch _:Err ->
```
I suppose - no.

robertoaloi replied:
Why not? It would be good to have the stacktrace available in the logs.

vkatsuba commented on apps/els_lsp/src/els_sheldon_diagnostics.erl:
```diff
@@ -0,0 +1,114 @@
+    RelFile ->
+      try
+        RegEx = "[_@./#&+-=*]",
+        Mod = rebar3_sheldon_ast,
```
Because we don't run `sheldon` together with the project but only when spellcheck diagnostics is running since it scares the `dialyzer` - I decided to use this approach.

vkatsuba replied:
Fixed.

vkatsuba commented on elvis.config:
```diff
@@ -34,6 +34,7 @@
                                                                         , els_tcp
                                                                         , els_dap_test_utils
                                                                         , els_test_utils
+                                                                        , els_sheldon_diagnostics
```
This is related to https://github.com/erlang-ls/erlang_ls/pull/1148#discussion_r760381453 discussion

vkatsuba replied:
Fixed

robertoaloi commented on apps/els_lsp/src/els_diagnostics.erl:
```diff
@@ -116,6 +118,7 @@ run_diagnostic(Uri, Id) ->
                     els_diagnostics_provider:notify(Diagnostics, self())
                 end
             },
+  ok = init_diagnostic(CbModule),
```
This would still be run every time a file is open or saved. I think we should run this once, during initialization.

vkatsuba replied:
Could you help me find better place?

robertoaloi replied:
As discussed via Slack, this could be implemented the `els_general_provider` when an initialization request is received or as part of the diagnostics bootstrap. I can have a deeper look at the code to see where to hook this exactly.

vkatsuba replied:
Done. But moved into `els_diagnostics_provider`.

robertoaloi commented on apps/els_lsp/src/els_sheldon_diagnostics.erl:
```diff
@@ -0,0 +1,119 @@
+%%==============================================================================
+%% Compiler diagnostics
```
Compiler -> Sheldon

robertoaloi commented on apps/els_lsp/src/els_sheldon_diagnostics.erl:
```diff
@@ -0,0 +1,119 @@
+
+-spec init() -> ok.
+init() ->
+  application:ensure_all_started(sheldon),
```
Maybe add a comment that this is executed here because sheldon has a long start time (something we should look into) and therefore we only do it if sheldon diagnostics are enabled?

robertoaloi commented on apps/els_lsp/src/els_sheldon_diagnostics.erl:
```diff
@@ -0,0 +1,119 @@
+    {error, not_relative} -> [];
+    RelFile ->
+      try
+        RegEx = "[_@./#&+-=*]",
```
Not being familiar with Sheldon, can you explain this regex?

vkatsuba replied:
This regex provide logic for rebar plugin for skip lines in source code, eg for `elvis.config` etc.

robertoaloi reviewed (commented):
Looks good, I left a couple of final comments. After that, we should be good to go.

robertoaloi commented:
Hi @vkatsuba ! I gave this a try, but after enabling the `sheldon` diagnostics backend via the `erlang_ls.config` file I see some crashes:

1. The `sheldon` functions are `undef`. I suspect the app is not included in the escript correctly (it probably misses in the `.app.src`
2. The `sheldon/priv/lang` folder required by the backend also doesn't seem to be available, so even after patching 1, the `sheldon_sup` crashes.

Also, it looks like sheldon comes with a 5MB english dictionary. Including it would make Erlang LS more than double its size. 
I wonder if, instead, we shouldn't provide the user the ability to point to a custom dictionary via the config file.
Finally, I suspect we will have to move the `ast` part from the rebar plugin to `sheldon` itself. The fact is that, by adding `rebar3_sheldon` as a dependency in the `.app.src` causes some rebar3 logic to trigger. In our internal project (1.5 million lines of code) this seems to trigger compilation for the whole project, which we cannot afford.

vkatsuba commented:
Hi @robertoaloi!

> The sheldon functions are undef. I suspect the app is not included in the escript correctly (it probably misses in the .app.src dependency chain.

We cannot add `sheldon` to `.app.src` because `sheldon` will try run and load and store dictionary by default. While it is expected that Sheldon can only start at a certain moment. Most likely you need to extend the plugin and add an API to run Sheldon separately.

> The sheldon/priv/lang folder required by the backend also doesn't seem to be available, so even after patching 1, the sheldon_sup crashes.

This folder is a part of `sheldon`

> Finally, I suspect we will have to move the ast part from the rebar plugin to sheldon itself.

If we do it, it will be no make sense to integrate `rebar3_sheldon`.

>  this seems to trigger compilation for the whole project, which we cannot afford.

The `rebar3_sheldon` works mostly the same as [rebar3_format](https://github.com/AdRoll/rebar3_format) which is also added as a dependency to the current project. I changed the API on purpose of `rebar3_sheldon` for current integration for call functions for analysis of AST.

___

Collect all this points - we should close this PR - correct?

robertoaloi commented:
So @vkatsuba if this doesn't work out of the box, how are you managing to get the diagnostics yourself?

vkatsuba commented:
@robertoaloi I was focused on common tests only and didn't try it in some IDE.
